
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>cyvcf2 &#8212; cyvcf2 0.30.18 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="cyvcf2 API" href="docstrings.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="cyvcf2">
<h1>cyvcf2<a class="headerlink" href="#cyvcf2" title="Permalink to this heading">¶</a></h1>
<p>cyvcf2 is a fast VCF parser for python (2 and 3) released under the MIT license.
It is a <cite>cython</cite> wrapper for <a class="reference external" href="https://github.com/samtools/htslib/">htslib</a> developed in
the <a class="reference external" href="http://quinlanlab.org/">Quinlan lab</a>.</p>
<p>See the <a class="reference internal" href="docstrings.html#api"><span class="std std-ref">cyvcf2 API</span></a> for detailed documentation, but the most common usage is summarized in the snippet below:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cyvcf2</span> <span class="kn">import</span> <span class="n">VCF</span>

<span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">VCF</span><span class="p">(</span><span class="s1">&#39;some.vcf.gz&#39;</span><span class="p">):</span> <span class="c1"># or VCF(&#39;some.bcf&#39;)</span>
    <span class="n">variant</span><span class="o">.</span><span class="n">REF</span><span class="p">,</span> <span class="n">variant</span><span class="o">.</span><span class="n">ALT</span> <span class="c1"># e.g. REF=&#39;A&#39;, ALT=[&#39;C&#39;, &#39;T&#39;]</span>

    <span class="n">variant</span><span class="o">.</span><span class="n">CHROM</span><span class="p">,</span> <span class="n">variant</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">variant</span><span class="o">.</span><span class="n">end</span><span class="p">,</span> <span class="n">variant</span><span class="o">.</span><span class="n">ID</span><span class="p">,</span> \
                <span class="n">variant</span><span class="o">.</span><span class="n">FILTER</span><span class="p">,</span> <span class="n">variant</span><span class="o">.</span><span class="n">QUAL</span>

    <span class="c1"># numpy arrays of specific things we pull from the sample fields.</span>
    <span class="c1"># gt_types is array of 0,1,2,3==HOM_REF, HET, UNKNOWN, HOM_ALT</span>
    <span class="n">variant</span><span class="o">.</span><span class="n">gt_types</span><span class="p">,</span> <span class="n">variant</span><span class="o">.</span><span class="n">gt_ref_depths</span><span class="p">,</span> <span class="n">variant</span><span class="o">.</span><span class="n">gt_alt_depths</span> <span class="c1"># numpy arrays</span>
    <span class="n">variant</span><span class="o">.</span><span class="n">gt_phases</span><span class="p">,</span> <span class="n">variant</span><span class="o">.</span><span class="n">gt_quals</span><span class="p">,</span> <span class="n">variant</span><span class="o">.</span><span class="n">gt_bases</span> <span class="c1"># numpy array</span>


    <span class="c1">## INFO Field.</span>
    <span class="c1">## extract from the info field by it&#39;s name:</span>
    <span class="n">variant</span><span class="o">.</span><span class="n">INFO</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;DP&#39;</span><span class="p">)</span> <span class="c1"># int</span>
    <span class="n">variant</span><span class="o">.</span><span class="n">INFO</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;FS&#39;</span><span class="p">)</span> <span class="c1"># float</span>
    <span class="n">variant</span><span class="o">.</span><span class="n">INFO</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;AC&#39;</span><span class="p">)</span> <span class="c1"># float</span>

    <span class="c1"># convert back to a string.</span>
    <span class="nb">str</span><span class="p">(</span><span class="n">variant</span><span class="p">)</span>


    <span class="c1">## per-sample info...</span>
    <span class="c1"># Get a numpy array of the depth per sample:</span>
    <span class="n">dp</span> <span class="o">=</span> <span class="n">variant</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s1">&#39;DP&#39;</span><span class="p">)</span>
    <span class="c1"># or of any other format field:</span>
    <span class="n">sb</span> <span class="o">=</span> <span class="n">variant</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s1">&#39;SB&#39;</span><span class="p">)</span>
    <span class="k">assert</span> <span class="n">sb</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="n">n_samples</span><span class="p">,</span> <span class="mi">4</span><span class="p">)</span> <span class="c1"># 4-values per</span>

<span class="c1"># to do a region-query:</span>

<span class="n">vcf</span> <span class="o">=</span> <span class="n">VCF</span><span class="p">(</span><span class="s1">&#39;some.vcf.gz&#39;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">vcf</span><span class="p">(</span><span class="s1">&#39;11:435345-556565&#39;</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">v</span><span class="o">.</span><span class="n">INFO</span><span class="p">[</span><span class="s2">&quot;AF&quot;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mf">0.1</span><span class="p">:</span> <span class="k">continue</span>
    <span class="nb">print</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">v</span><span class="p">))</span>

    <span class="c1"># single sample of 0|1 in vcf becomes [[0, 1, True]]</span>
    <span class="c1"># 2 samples of 0/0 and 1|1 would be [[0, 0, False], [1, 1, True]]</span>
    <span class="nb">print</span> <span class="n">v</span><span class="o">.</span><span class="n">genotypes</span>
</pre></div>
</div>
</section>
<section id="modifying-existing-records">
<h1>Modifying Existing Records<a class="headerlink" href="#modifying-existing-records" title="Permalink to this heading">¶</a></h1>
<p><cite>cyvcf2</cite> is optimized for fast reading and extraction from existing files.
However, it also offers some means of modifying existing VCFs. Here, we
show an example of how to modify the INFO field at a locus to annotate
variants with the genes that they overlap.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cyvcf2</span> <span class="kn">import</span> <span class="n">VCF</span><span class="p">,</span> <span class="n">Writer</span>
<span class="n">vcf</span> <span class="o">=</span> <span class="n">VCF</span><span class="p">(</span><span class="n">VCF_PATH</span><span class="p">)</span>
<span class="c1"># adjust the header to contain the new field</span>
<span class="c1"># the keys &#39;ID&#39;, &#39;Description&#39;, &#39;Type&#39;, and &#39;Number&#39; are required.</span>
<span class="n">vcf</span><span class="o">.</span><span class="n">add_info_to_header</span><span class="p">({</span><span class="s1">&#39;ID&#39;</span><span class="p">:</span> <span class="s1">&#39;gene&#39;</span><span class="p">,</span> <span class="s1">&#39;Description&#39;</span><span class="p">:</span> <span class="s1">&#39;overlapping gene&#39;</span><span class="p">,</span>
    <span class="s1">&#39;Type&#39;</span><span class="p">:</span><span class="s1">&#39;Character&#39;</span><span class="p">,</span> <span class="s1">&#39;Number&#39;</span><span class="p">:</span> <span class="s1">&#39;1&#39;</span><span class="p">})</span>

<span class="c1"># create a new vcf Writer using the input vcf as a template.</span>
<span class="n">fname</span> <span class="o">=</span> <span class="s2">&quot;out.vcf&quot;</span>
<span class="n">w</span> <span class="o">=</span> <span class="n">Writer</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span> <span class="n">vcf</span><span class="p">)</span>

<span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">vcf</span><span class="p">:</span>
    <span class="c1"># The get_gene_intersections function is not shown.</span>
    <span class="c1"># This could be any operation to find intersections</span>
    <span class="c1"># or any manipulation required by the user.</span>
    <span class="n">genes</span> <span class="o">=</span> <span class="n">get_gene_intersections</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">genes</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">v</span><span class="o">.</span><span class="n">INFO</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span>
    <span class="n">w</span><span class="o">.</span><span class="n">write_record</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>

<span class="n">w</span><span class="o">.</span><span class="n">close</span><span class="p">();</span> <span class="n">vcf</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>More info on writing vcfs can be found <a class="reference internal" href="writing.html"><span class="doc">here</span></a></p>
</section>
<section id="setting-genotyping-strictness">
<h1>Setting Genotyping Strictness<a class="headerlink" href="#setting-genotyping-strictness" title="Permalink to this heading">¶</a></h1>
<p>By default, genotypes containing a single missing allele (e.g. genotypes such
as <code class="docutils literal notranslate"><span class="pre">0/.</span></code>, <code class="docutils literal notranslate"><span class="pre">./0</span></code>, <code class="docutils literal notranslate"><span class="pre">1/.</span></code>, or <code class="docutils literal notranslate"><span class="pre">./1</span></code>) are classified as heterozygous (“HET”
or id: 1) instead of “UNKNOWN” (or id: 2).  A case can be made for either
classification on these partially missing genotypes depending on the situation.</p>
<p>A better way to explicitly control the genotype classification for these cases,
is to use the “strict genotype”, or <code class="docutils literal notranslate"><span class="pre">strict_gt</span></code>, feature.  Here’s an example
usage:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cyvcf2</span> <span class="kn">import</span> <span class="n">VCF</span>
<span class="n">vcf</span> <span class="o">=</span> <span class="n">VCF</span><span class="p">(</span><span class="s2">&quot;/path/to/vcf/file&quot;</span><span class="p">,</span> <span class="n">strict_gt</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">vcf</span><span class="p">:</span>
    <span class="c1"># do something</span>
</pre></div>
</div>
<p>When the <code class="docutils literal notranslate"><span class="pre">strict_gt</span></code> flag is enabled, <cite>cyvcf2</cite> will treat any genotype
containing a missing allele (containing a ‘.’) as an UNKNOWN genotype;
otherwise, genotypes like <code class="docutils literal notranslate"><span class="pre">0/.</span></code>, <code class="docutils literal notranslate"><span class="pre">./0</span></code>, <code class="docutils literal notranslate"><span class="pre">1/.</span></code>, or <code class="docutils literal notranslate"><span class="pre">./1</span></code> will be
classified as heterozygous (“HET”).</p>
<p>The “strict genotype” feature is not enabled by default (i.e. <code class="docutils literal notranslate"><span class="pre">strict_gt</span></code> is
set to <cite>False</cite> by default) to preserve backwards compatibility.</p>
</section>
<section id="installation">
<h1>Installation<a class="headerlink" href="#installation" title="Permalink to this heading">¶</a></h1>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip install cyvcf2
</pre></div>
</div>
<p>or via bioconda.</p>
</section>
<section id="testing">
<h1>Testing<a class="headerlink" href="#testing" title="Permalink to this heading">¶</a></h1>
<p>Install <cite>pytest</cite>, then tests can be run with:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pytest
</pre></div>
</div>
</section>
<section id="known-limitations">
<h1>Known Limitations<a class="headerlink" href="#known-limitations" title="Permalink to this heading">¶</a></h1>
<ul class="simple">
<li><p><cite>cyvcf2</cite> currently does not support reading VCFs encoded with UTF-8 with non-ASCII characters in the contents of string-typed FORMAT fields.</p></li>
</ul>
<p>For limitations on writing VCFs, see <a class="reference internal" href="writing.html#limitations-with-writing"><span class="std std-ref">here</span></a></p>
</section>
<section id="see-also">
<h1>See Also<a class="headerlink" href="#see-also" title="Permalink to this heading">¶</a></h1>
<p>Pysam also <a class="reference external" href="https://github.com/pysam-developers/pysam/blob/master/pysam/cbcf.pyx">has a cython wrapper to htslib</a> and one block of code here is taken directly from that library. But, the optimizations that we want for gemini are very specific so we have chosen to create a separate project.</p>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="docstrings.html">cyvcf2 API</a></li>
<li class="toctree-l1"><a class="reference internal" href="writing.html">Modifying Existing VCFs</a></li>
</ul>
</div>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="#">cyvcf2</a></h1>








<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="docstrings.html">cyvcf2 API</a></li>
<li class="toctree-l1"><a class="reference internal" href="writing.html">Modifying Existing VCFs</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="#">Documentation overview</a><ul>
      <li>Next: <a href="docstrings.html" title="next chapter">cyvcf2 API</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2016..2020, Brent Pedersen.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 5.3.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/index.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>